### PR TITLE
[move-prover] Bump up the version number of the Prover dependencies

### PR DESCRIPTION
**Pull request by junkil-park**
- Updated the Z3 version to the latest one.
- Updated the Boogie version to latest one that is compatible with Prover.
  - Incompatibility issue filed in the Boogie repo: https://github.com/boogie-org/boogie/issues/609
- Fixed a few bugs in `boogie_wrapper`.

## Motivation

To bump up the version number of Prover's deps.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/move-language/move/blob/main/CONTRIBUTING.md#developer-workflow)?

yes

## Test Plan
cargo test -p move-prover

### CODE REVIEW

File: language/move-prover/boogie-backend/src/boogie_wrapper.rs [hunk 1]
```diff
@@ -1682,7 +1682,7 @@ impl<'s> ModelParser<'s> {
 
     fn parse_key(&mut self) -> Result<ModelValue, ModelParseError> {
         let mut comps = vec![];
-        while !self.looking_at("->") {
+        while !self.looking_at("->") && self.at < self.input.len() {
```
Comment (meng-xu-cs) [inline language/move-prover/boogie-backend/src/boogie_wrapper.rs:1685]: I am running Ubuntu 20.04.4 LTS and I realize that my prover tests pass without this line of change...

File: scripts/dev_setup.sh [hunk 2]
```diff
@@ -15,10 +15,10 @@
 # fast fail.
 set -eo pipefail
 
-Z3_VERSION=4.8.13
+Z3_VERSION=4.10.1
```
Comment (meng-xu-cs) [inline scripts/dev_setup.sh:18]: Also, Z3 [4.10.2](https://github.com/Z3Prover/z3/releases/tag/z3-4.10.2) is released as well.

File: language/move-prover/boogie-backend/src/boogie_wrapper.rs (hunk 1 above)
Comment (meng-xu-cs) [reply language/move-prover/boogie-backend/src/boogie_wrapper.rs:1685]: So I still don't quite understand what this line is doing...

File: scripts/dev_setup.sh (hunk 2 above)
Comment (junkil-park) [reply scripts/dev_setup.sh]: Done.

File: language/move-prover/boogie-backend/src/boogie_wrapper.rs (hunk 1 above)
Comment (junkil-park) [reply language/move-prover/boogie-backend/src/boogie_wrapper.rs:1685]: Sud0u53r at our Discord channel points out that this is a bug which causes Prover fall into an infinite loop when `boogie_wrapper` tries to parse an incomplete map expression with no value part (like `var ->`). This is a clear bug that needs to be fixed.

Comment (junkil-park) [reply language/move-prover/boogie-backend/src/boogie_wrapper.rs:1685]: > I am running Ubuntu 20.04.4 LTS and I realize that my prover tests pass without this line of change...
If you upgrade Boogie version to 2.14.0, Boogie will start producing incomplete expressions like (`$abort_code ->`). If you run Prover again without this fix, Prover will not terminate due to the infinite loop bug. If you run Prover with this fix, Prover will terminate, but complain about such incomplete expressions.

I am trying to be confirmed whether such incomplete expressions are the expected behavior of Boogie or a bug. In the meantime, we can safely upgrade and use the Boogie version of 2.13.4.

Comment (meng-xu-cs) [reply language/move-prover/boogie-backend/src/boogie_wrapper.rs:1685]: I see. Fair enough

**Review by meng-xu-cs — APPROVED** [review]

**Pull request merged**